src/flamegraph/color/palette_map.rs

```diff
@@ -5,67 +5,96 @@ use std::fs::OpenOptions;
 use std::io;
 use std::io::BufRead;
 use std::io::BufReader;
-use std::io::Write;
 use std::path::Path;
 use std::str::FromStr;
 
-pub(crate) struct PaletteMap<'a>(HashMap<Cow<'a, str>, (u8, u8, u8)>);
+/// Mapping of the association between a function name and the color used when drawing information
+/// from this function.
+#[derive(Default)]
+pub struct PaletteMap<'a>(HashMap<Cow<'a, str>, (u8, u8, u8)>);
 
 impl<'a> PaletteMap<'a> {
-    pub fn load(file: &str) -> quick_xml::Result<Self> {
+    /// Returns the color value corresponding to the given function name.
+    pub fn get(&self, func: &str) -> Option<(u8, u8, u8)> {
+        self.0.get(func).cloned()
+    }
+
+    /// Inserts a function name/color pair in the map.
+    pub fn insert(&mut self, func: &'a str, color: (u8, u8, u8)) -> Option<(u8, u8, u8)> {
+        self.0.insert(Cow::from(func), color)
+    }
+
+    /// Provides an iterator over the elements of the map.
+    pub fn iter(&self) -> impl Iterator<Item = (&str, (u8, u8, u8))> {
+        self.0.iter().map(|(func, color)| (func.as_ref(), *color))
+    }
+
+    /// Builds a mapping based on the inputs given by the reader.
+    ///
+    /// The reader should provide name/color pairs as text input, each pair separated by a line
+    /// separator.
+    ///
+    /// Each line should follow the format: NAME->rgb(RED, GREEN, BLUE)
+    /// where NAME is the function name, and RED, GREEN, BLUE integer values between 0 and 255
+    /// included.
+    ///
+    /// This function will return an [`std::io::Error`] if the input is not correctly formatted.
+    pub fn from_stream(reader: &mut dyn io::Read) -> io::Result<Self> {
         let mut map = HashMap::default();
-        let path = Path::new(file);
+        let reader = BufReader::new(reader);
```

Comment:
Maybe we should instead just take a `&mut dyn io::BufRead` directly?

Reply:
I am not sure of the implications of such a change. Do we do not want to provide the user with the most generic interface ? Or is there an advantage for the user to provide a different `BufRead` implementation that the one we use ?

Reply:
It means that if they already have something that implements `BufRead` (like `Cursor` or a `BufReader`) we don't add our own buffer on top of that. In some sense, switching to `BufRead` gives the user _more_ choice. They can always wrap it in a `BufReader`, and we don't particularly care!